lib/luaxtable.c

```diff
@@ -43,37 +47,123 @@ typedef struct luaxtable_s {
 {
-	return 0;
+	lunatik_require(L, data);
+	lunatik_object_t *data = lunatik_checknull(L, luadata_new(ptr, size, sleep));
+	lunatik_cloneobject(L, data);
 }
 
-static void luaxtable_match_destroy(const struct xt_mtdtor_param *par)
+#define luaxtable_getrcu(xtable, hook, key, keylen)			\
+do {									\
+	lunatik_object_t *obj = luarcu_gettable(xtable_##hook##hooks, key, keylen);	\
+	if (obj == NULL) {						\
+		pr_err("could not find table for %s\n", key);		\
+		return -1;						\
+	}								\
+	xtable = (luaxtable_t *)obj->private;				\
+} while(0)
+
+#define luaxtable_setrcu(object, hook, key, keylen)			\
+do {									\
+	if (luarcu_settable(xtable_##hook##hooks, key, keylen, object) != 0) {	\
+		luaL_error(L, "unable to set table");			\
+		return -1;						\
+	}								\
+} while(0)
+
+
+static int luaxtable_invoke(lua_State *L, luaxtable_t *xtable, struct sk_buff *skb, const char *op, int nret)
 {
+	int base = lua_gettop(L);
+	int ret = -ENXIO;
+	int nargs = 0;
+
+	if (lunatik_getregistry(L, xtable) != LUA_TTABLE) {
+		pr_err("could not find ops table\n");
+		goto err;
+	}
+
+	if (lua_getfield(L, -1, op) != LUA_TFUNCTION) {
+		pr_err("%s isn't defined\n", op);
+		goto err;
+	}
+
+	if (skb != NULL ){
+		luaxtable_newdata(L, skb, sizeof(struct sk_buff), false);
+		nargs = 1;
+	}
+
+	if (lua_pcall(L, nargs, nret, 0) != LUA_OK) {
+		pr_err("%s error: %s\n", op, lua_tostring(L, -1));
+		ret = -ECANCELED;
+		goto err;
+	}
+
+	return nret > 0 ? luaL_optinteger(L, -1, 0) : 0;
+err:
+	lua_settop(L, base); /* pop everything, including args */
+	return ret;
 }
 
-static unsigned int luaxtable_target(struct sk_buff *skb, const struct xt_action_param *par)
-{
-	return NF_ACCEPT;
+#define LUAXTABLE_HOOK_CB(T, U, V, hook, huk) 				\
+static T luaxtable_##hook(U skb, V par)					\
+{									\
+	const luaxtable_info_t *info = (const luaxtable_info_t *)par->huk##info;	\
+	luaxtable_t *xtable = info->data;				\
+	int ret = xtable->fallback;					\
+									\
+	/* TODO: handle const type in luadata */			\
+	lunatik_run(xtable->runtime, luaxtable_invoke, ret, xtable, (struct sk_buff *)skb, #hook, 1);	\
+	if (ret < 0)							\
+		pr_err("error in running " #hook " lua handler : %d\n", ret);	\
+	return ret < 0 ? xtable->fallback : ret;					\
 }
 
-static int luaxtable_target_check(const struct xt_tgchk_param *par)
-{
-	return 0;
+#define LUAXTABLE_CHECK_CB(hook, hk, huk, HOOK)					\
+static int luaxtable_##hook##_check(const struct xt_##hk##chk_param *par)	\
+{										\
+	int ret = -EINVAL;							\
+										\
+	luaxtable_t *xtable;							\
+	luaxtable_getrcu(xtable, hook, par->hook->name, XT_EXTENSION_MAXNAMELEN);	\
+	luaxtable_info_t *info = (luaxtable_info_t *)par->huk##info; 		\
+	info->data = xtable;							\
+										\
+	lunatik_run(xtable->runtime, luaxtable_invoke, ret, xtable, NULL, "checkentry", 1);	\
+	if (ret < 0)								\
+		pr_err("error in running 'checkentry' lua handler : %d\n", ret);	\
```

Comment:
we can move these prints to `invoke`, right? we have the `op` string there.. it will clean up a bit your macros..